src/beast/beast/http/README.md

```diff
@@ -0,0 +1,401 @@
+}
+```
+
+Messages can be read from the network and parsed into a `parsed_message` object,
+which extends the `message` by adding parse-specific metadata such as the
+keep-alive which is context sensitive (depending on the HTTP version for
+example). When preparing a response for sending, `prepare` must be called with
+an additional parameter, the corresponding parsed request. The implementation
+inspects the contents of the request to set dependent fields of the response.
+This example reads a message, builds a response, and sends it.
+```C++
+void handle_connection(ip::tcp::socket& sock)
+{
+    parsed_request<string_body> req;
+    read(sock, req);
+    response<string_body> resp;
+    ...
+    write(sock, prepare(resp, req));
+}
+```
+
+## Modeling the HTTP message
+
+All HTTP messages are modeled using this base class template:
+```C++
+template<bool isRequest, class Body, class Allocator>
+class message
+{
+    ...
+    typename Body::value_type body;
+}
+
+template<class Body, class Allocator>
+using request = message<true, Body, Allocator>;
+
+template<class Body, class Allocator>
+using response = message<false, Body, Allocator>;
+```
+
+The template argument `isRequest` is `true` for HTTP requests and `false`
+for HTTP responses, allowing functions to be overloaded or constrained
+based on the type of message they want to be passed.
+
+The `Body` template argument controls the method used to store information
+necessary for receiving or sending the body, as well as providing customizations
+for the actual writing or parsing process. The customizations are used by the
+implementation to perform the `read`, `write`, `async_read`, and `async_write`
+operations on messages.
+
+Beast.HTTP offers `empty_body`, `string_body`, and `streambuf_body` as common
+choices for the `Body` template argument. User-defined objects that meet the `Body`
+requirements may be implemented for custom implementation strategies.
+
+*Note:* Definitions for member functions associated with a `Body` and the types
+it defines should typically be declared inline so they become part of the calling
+code.
+
+## Concepts
+
+### `Body`
+
+Requirements:
+
+`req` denotes any instance of `prepared_request`.<br>
+`resp` denotes any instance of `prepared_response`.<br>
+`preq` denotes any instance of `parsed_request`.<br>
+
+ expression                 | return | type assertion/note/pre/post-condition
+:-------------------------- |:------ |:--------------------------------------
+`Body::value_type`          |        | The type of the `message::body` member.
+`Body::value_type{}`        |        | `DefaultConstructible`
+`Body::reader`              |        | A type meeting the requirements of `Reader`
+`Body::writer`              |        | A type meeting the requirements of `SinglePassWriter` or `MultiPassWriter`
+`Body::prepare(req)`        |        | Prepare `req` for serialization
+`Body::prepare(resp, preq)` |        | Prepare `resp` for serialization
+
```

Comment:
I don't think so. The Body should get the prepared version of the message. There are probably some warts with this interface. Its a crucial part of the design of the library. I could use some help with it.